src/lexer.cc

```diff
@@ -0,0 +1,139 @@
+/*
+   Copyright 2019 Simon (psyomn) Symeonidis
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+#include "bk/lexer.h"
+#include "bk/token.h"
+
+#include <filesystem>
+#include <sstream>
+#include <fstream>
+
+namespace bk {
+static bool isPunct(char c) {
+  return
+    c == '{' || c == '}' ||
+    c == '.' || c == ',' ||
+    c == ';' || c == ':' ||
+    c == '\'' || c == '"' ||
+    c == '!' || c == '?' ||
+    c == '-' ||
+    c == '(' || c == ')' ||
+    c == '[' || c == ']' ||
+    c == '<' || c == '>';
+}
+
+static bool isSymbol(char c) {
+  return
+    c == '@' || c == '#' || c == '$' || c == '^' ||
+    c == '&' || c == '*' || c == '|' || c == '\\' ||
+    c == '/';
+}
+
+static bool isNewline(char c) {
+  return c == '\n';
+}
+
+static bool isWhitespace(char c) {
+  // ignore chars
+  return c == ' ' || c == '\t' || c == '\r';
+}
+
+Lexer::Lexer(const std::string& path) :
+  mLineNum(1),
+  mAcceptExtension(".md") {
+  mRoot = path;
+}
+
+Lexer::~Lexer() {}
+
+void Lexer::process() {
+  namespace fs = std::filesystem;
+
+  for (auto& el : fs::recursive_directory_iterator(mRoot))
+    if (el.path().extension() == mAcceptExtension) mFiles.push_back(el.path());
+
+  for (auto& path : mFiles) tokenizeFile(path);
+}
+
+void Lexer::tokenizeFile(const std::string& filename) {
+  std::ifstream file(filename, std::ifstream::in);
+  std::stringstream ss;
+  std::vector<bk::Token> tokens;
+  std::size_t linecount = 0;
+  char c;
+
+  while (file.get(c)) {
+    if (isWhitespace(c)) continue;
+
+    if (isNewline(c)) {
+      linecount++;
+      bk::Token tok(std::string(""), linecount);
+      tok.type(Token::Type::newline);
+      tokens.push_back(tok);
+
+      ss.str(std::string());
```

Comment:
If you want, C++11 allows the following syntax:
```suggestion
      ss.str({});
```
Basically, they added the ability to construct object via the curly-brace syntax that you might be familliar with in C. The type to be constructed is inferred by the context.